### ISSUE TITLE
A widget's Drag and Drop does not work when dragging an url from Chrome on Linux

### ISSUE THREAD

**Issue by ggarra13** [fltk/fltk#988]
**Describe the bug**
Trying to drag an URL from Chrome onto a widget that takes DnD (FL_PASTE) events, does not work on Linux.
Firefox works fine.

Problem is not present on macOS nor Windows. 

**To Reproduce**
[main.cpp.txt](https://github.com/user-attachments/files/15597455/main.cpp.txt)

1. fltk-config --compile main.cpp
2. Run "main"
3. Open Google chrome, go to any URL, like www.google.com.  Select the full URL in the URI bar.  Drag it onto DROP ON ME! widget.
4. FL_PASTE nor any of the DND events are called.

**Expected behavior**
Drag and drop of the URI text.

**FLTK Version**
1.4 master

**Operating System / Platform:**
Linux Ubuntu 22.04.4 LTS
Linux Rocky Linux 8.9

**Linux/Unix Runtime, if applicable:**
 - Wayland
 - X11

**Additional context**
The test program clipboard.cxx does display the text as stored in Fl::clipboard_plain_text.

**Comment by ggarra13** [comment]
It seems the XdndAware property is not set right (old version?).  This code seems to fix it:

[main.cpp.txt](https://github.com/user-attachments/files/15605526/main.cpp.txt)

**Comment by ggarra13** [comment]
This patch seems to fix the problem:

[Fl_x.patch.txt](https://github.com/user-attachments/files/15605594/Fl_x.patch.txt)

**Comment by ggarra13** [comment]
On Wayland, the bug seems to be with Chrome itself, as it will not drag its contents to the Gnome console, Emacs nor any FLTK application.

If you agree that my X11 patch is good, you can commit it and close the issue.

**Comment by Albrecht-S** [comment]
@ggarra13 Thanks for the report. I confirm the issue using Google Chromium in my case. I applied your patch and noticed that it really used version 5 of the XDND protocol and that no XdndEnter message is sent if we use the old code which allows only protocol version 4.

> If you agree that my X11 patch is good, you can commit it and close the issue.

I'm currently checking the changes needed in XDND protocol version 5. I'll let you know if the patch is sufficient, or I'll add the required "bits".

FTR: See ["Drag-and-Drop Protocol for the X Window System"](https://www.freedesktop.org/wiki/Specifications/XDND/), look for "version 5", particularly in the description of the `XdndFinished` message.

**Comment by ggarra13** [comment]
Also, I noticed:

  fl_XdndEnter          = XInternAtom(d, "XdndEnter",           0);
  
is defined twice in the initialization code.

**Comment by Albrecht-S** [comment]
Yup, thanks, I noticed this too. I'll remove one of the definitions.

**Comment by ggarra13** [comment]
I read about XdndFinished, and my interpretation of it is in this patch for review:

[Fl_x.patch.txt](https://github.com/user-attachments/files/15614559/Fl_x.patch.txt)

**Comment by Albrecht-S** [comment]
Thanks for your updated patch. I don't think that this is correct WRT our current code.

Currently we send `XdndFinished` **after** the `FL_PASTE` event in [line 1467](https://github.com/fltk/fltk/blob/651d9f3e7960fc17d0ab1b27830d160fac1be132/src/Fl_x.cxx#L1467) so we (can) know if the data was accepted (`retval`). Your code would add another `XdndFinished` message before we know if the payload was accepted.

The subsequent `XdndFinished` message in [line 1479](https://github.com/fltk/fltk/blob/651d9f3e7960fc17d0ab1b27830d160fac1be132/src/Fl_x.cxx#L1479) should IMHO transmit the return value (`retval`) of the `FL_PASTE` event. I'm still investigating...

See also [this comment](https://github.com/fltk/fltk/blob/651d9f3e7960fc17d0ab1b27830d160fac1be132/src/Fl_x.cxx#L1474) in the code:
```
It is not clear if this has to be delayed until now or if it can be done
immediately after calling XConvertSelection.
```
I believe that this is clear now since `XDND` protocol version 5 allows to send the return value, hence we must delay sending `XdndFinished` until after the `FL_PASTE` event.

Note that I'm not an X11 expert, this is just how I interpret the docs.

Comments, anybody?

**Comment by ggarra13** [comment]
You are absolutely correct.   I forgot to clear fl_dnd_source_window which would avoid the double sending, but since the FLTK docs specify that FL_PASTE is sent right AFTER FL_DND_RELEASE, waiting till FL_PASTE is the correct approach.

I am attaching a new version of my patch for review:

[Fl_x.cxx_v2.patch.txt](https://github.com/user-attachments/files/15654446/Fl_x.cxx_v2.patch.txt)

**Comment by Albrecht-S** [comment]
Thanks, technically your patch is almost identical to my work, but I added some more docs and I changed a tiny bit differently to make sure we set only the relevant bit:
```diff
+    if (fl_xevent->xselection.property == XA_SECONDARY && fl_dnd_source_window) {
+      fl_sendClientMessage(fl_dnd_source_window,            // send to window
+                           fl_XdndFinished,                 // XdndFinished message
+                           fl_xevent->xselection.requestor, // data.l[0] target window
+                           retval ? 1 : 0,                  // data.l[1] Bit 0: 1 = success
+                           retval ? fl_dnd_action : None);  // data.l[2] action performed
```

I'm still looking for some other aspects, but so far I'm pretty sure that this version can be committed.

**Comment by ManoloFLTK** [comment]
FTR: I've been looking at what happens when doing DnD from Chromium (which is an X11 client) to a Wayland FLTK app. This [blog post](https://emersion.fr/blog/2020/wayland-clipboard-drag-and-drop/) describes how DnD works under Wayland. None of the two initial events that a DnD destination app should receive are received:
1) the destination client needs to listen to wl_data_device.data_offer events to handle new data offers.
2) When the user has initiated a drag-and-drop operation and moves the cursor over one of the destination client’s surfaces, a wl_data_device.enter event will be sent by the compositor. 

Therefore, I see nothing to be done in FLTK Wayland source code to fix that. The error seems to be somewhere in Chromium or in how the Wayland compositor (Mutter here) transforms an X11 DnD into a Wayland DnD.

**Comment by ManoloFLTK** [comment]
There's a good workaround, though, to work with Wayland: `chromium` can be made a Wayland client launching it as follows:
`chromium --ozone-platform-hint=wayland &`

With this, DnD from the chromium URL field to an FLTK Wayland client app works both with gnome/Mutter and KDE/Plasma.

**Comment by ggarra13** [comment]
@ManoloFLTK Thank you for looking into this and for fixing the Wayland transparency bug.   I tested chrome with the ozone platform hint and it is currently way too buggy under Ubuntu 22.04.4 LTS, at least.

**Comment by Albrecht-S** [comment]
I wrote:
> I'm still looking for some other aspects, but so far I'm pretty sure that this version can be committed.

As written above, the fix in commit 3cc5c090f9b24180db428d55dceffd0aa7c7228c seems to be good. I committed and pushed it.

The Wayland issues mentioned above are another issue, no matter whether they can be or have been fixed or not. If there is need to pursue this Wayland DND stuff further, please open another issue.

I'm closing this issue now since the main issue (X11 - DND) has been resolved. Thanks to Gonzalo (@ggarra13) for the report and the patch suggestions.
